File: pkg/sortname/sortname_test.go
package sortname

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestForTitle(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		// Basic article handling
		{
			name:     "The at beginning",
			input:    "The Hobbit",
			expected: "Hobbit, The",
		},
		{
			name:     "A at beginning",
			input:    "A Tale of Two Cities",
			expected: "Tale of Two Cities, A",
		},
		{
			name:     "An at beginning",
			input:    "An American Tragedy",
			expected: "American Tragedy, An",
		},

		// Case insensitivity
		{
			name:     "the lowercase",
			input:    "the hobbit",
			expected: "hobbit, the",
		},
		{
			name:     "THE uppercase",
			input:    "THE HOBBIT",
			expected: "HOBBIT, THE",
		},

		// No article
		{
			name:     "no article",
			input:    "Lord of the Rings",
			expected: "Lord of the Rings",
		},
		{
			name:     "article in middle only",
			input:    "Return of the King",
			expected: "Return of the King",
		},

		// Edge cases
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "whitespace only",
			input:    "   ",
			expected: "",
		},
		{
			name:     "just The",
			input:    "The",
			expected: "The",
		},
		{
			name:     "The with whitespace",
			input:    "The ",
			expected: "The",
		},
		{
			name:     "single word no article",
			input:    "Dune",
			expected: "Dune",
		},

		// Real world examples
		{
			name:     "The Lord of the Rings",
			input:    "The Lord of the Rings",
			expected: "Lord of the Rings, The",
		},
		{
			name:     "A Game of Thrones",
			input:    "A Game of Thrones",
			expected: "Game of Thrones, A",
		},
		{
			name:     "The Great Gatsby",
			input:    "The Great Gatsby",
			expected: "Great Gatsby, The",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := ForTitle(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestForPerson(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		// Basic name inversion
		{
			name:     "simple two-part name",
			input:    "Stephen King",
			expected: "King, Stephen",
		},
		{
			name:     "three-part name",
			input:    "Martin Luther King",
			expected: "King, Martin Luther",
		},

		// Generational suffixes (preserved)
		{
			name:     "Jr suffix",
			input:    "Robert Downey Jr.",
			expected: "Downey, Robert, Jr.",
		},
		{
			name:     "Jr without period",
			input:    "Robert Downey Jr",
			expected: "Downey, Robert, Jr",
		},
		{
			name:     "Sr suffix",
			input:    "John Smith Sr.",
			expected: "Smith, John, Sr.",
		},
		{
			name:     "III suffix",
			input:    "John Smith III",
			expected: "Smith, John, III",
		},
		{
			name:     "II suffix",
			input:    "Robert Kennedy II",
			expected: "Kennedy, Robert, II",
		},
		{
			name:     "Martin Luther King Jr",
			input:    "Martin Luther King Jr.",
			expected: "King, Martin Luther, Jr.",
		},

		// Academic suffixes (stripped)
		{
			name:     "PhD suffix",
			input:    "Jane Doe PhD",
			expected: "Doe, Jane",
		},
		{
			name:     "Ph.D. suffix",
			input:    "Jane Doe Ph.D.",
			expected: "Doe, Jane",
		},
		{
			name:     "MD suffix",
			input:    "John Smith MD",
			expected: "Smith, John",
		},
		{
			name:     "M.D. suffix",
			input:    "John Smith M.D.",
			expected: "Smith, John",
		},
		{
			name:     "PsyD suffix",
			input:    "Sarah Connor PsyD",
			expected: "Connor, Sarah",
		},
		{
			name:     "MBA suffix",
			input:    "Bob Jones MBA",
			expected: "Jones, Bob",
		},
		{
			name:     "multiple academic suffixes",
			input:    "John Doe MD PhD",
			expected: "Doe, John",
		},

		// Mixed generational and academic (generational preserved, academic stripped)
		{
			name:     "Jr and PhD",
			input:    "John Smith Jr. PhD",
			expected: "Smith, John, Jr.",
		},

		// Prefixes (stripped)
		{
			name:     "Dr prefix",
			input:    "Dr. Sarah Connor",
			expected: "Connor, Sarah",
		},
		{
			name:     "Dr without period",
			input:    "Dr Sarah Connor",
			expected: "Connor, Sarah",
		},
		{
			name:     "Mr prefix",
			input:    "Mr. John Smith",
			expected: "Smith, John",
		},
		{
			name:     "Mrs prefix",
			input:    "Mrs. Jane Doe",
			expected: "Doe, Jane",
		},
		{
			name:     "Prof prefix",
			input:    "Prof. Albert Einstein",
			expected: "Einstein, Albert",
		},
		{
			name:     "Sir prefix",
			input:    "Sir Isaac Newton",
			expected: "Newton, Isaac",
		},

		// Prefix and suffix combined
		{
			name:     "Dr with PhD",
			input:    "Dr. John Smith PhD",
			expected: "Smith, John",
		},
		{
			name:     "Prof with multiple degrees",
			input:    "Prof. Jane Doe MD PhD",
			expected: "Doe, Jane",
		},

		// Particles (moved to end)
		{
			name:     "van Beethoven",
			input:    "Ludwig van Beethoven",
			expected: "Beethoven, Ludwig van",
		},
		{
			name:     "von Neumann",
			input:    "John von Neumann",
			expected: "Neumann, John von",
		},
		{
			name:     "da Vinci",
			input:    "Leonardo da Vinci",
			expected: "Vinci, Leonardo da",
		},
		{
			name:     "de Gaulle",
			input:    "Charles de Gaulle",
			expected: "Gaulle, Charles de",
		},
		{
			name:     "del Toro",
			input:    "Guillermo del Toro",
			expected: "Toro, Guillermo del",
		},
		{
			name:     "van der Waals",
			input:    "Johannes van der Waals",
			expected: "Waals, Johannes van der",
		},

		// Particle with suffix
		{
			name:     "van with Jr",
			input:    "John van Smith Jr.",
			expected: "Smith, John van, Jr.",
		},

		// Edge cases
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "whitespace only",
			input:    "   ",
			expected: "",
		},
		{
			name:     "single name",
			input:    "Madonna",
			expected: "Madonna",
		},
		{
			name:     "single name with whitespace",
			input:    "  Cher  ",
			expected: "Cher",
		},

		// Real world examples
		{
			name:     "J.R.R. Tolkien",
			input:    "J.R.R. Tolkien",
			expected: "Tolkien, J.R.R.",
		},
		{
			name:     "George R.R. Martin",
			input:    "George R.R. Martin",
			expected: "Martin, George R.R.",
		},
		{
			name:     "H.P. Lovecraft",
			input:    "H.P. Lovecraft",
			expected: "Lovecraft, H.P.",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := ForPerson(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestIsPrefix(t *testing.T) {
	tests := []struct {
		word     string
		expected bool
	}{
		{"Dr.", true},
		{"Dr", true},
		{"dr.", true},
		{"DR", true},
		{"Mr.", true},
		{"Mrs.", true},
		{"Prof.", true},
		{"Sir", true},
		{"John", false},
		{"", false},
	}

	for _, tt := range tests {
		t.Run(tt.word, func(t *testing.T) {
			assert.Equal(t, tt.expected, isPrefix(tt.word))
		})
	}
}

func TestIsGenerationalSuffix(t *testing.T) {
	tests := []struct {
		word     string
		expected bool
	}{
		{"Jr.", true},
		{"Jr", true},
		{"jr.", true},
		{"JR", true},
		{"Sr.", true},
		{"III", true},
		{"iii", true},
		{"II", true},
		{"IV", true},
		{"V", true},
		{"Junior", true},
		{"Senior", true},
		{"PhD", false},
		{"John", false},
		{"", false},
	}

	for _, tt := range tests {
		t.Run(tt.word, func(t *testing.T) {
			assert.Equal(t, tt.expected, isGenerationalSuffix(tt.word))
		})
	}
}

func TestIsAcademicSuffix(t *testing.T) {
	tests := []struct {
		word     string
		expected bool
	}{
		{"PhD", true},
		{"Ph.D.", true},
		{"phd", true},
		{"MD", true},
		{"M.D.", true},
		{"PsyD", true},
		{"Psy.D.", true},
		{"MBA", true},
		{"Esq.", true},
		{"Jr.", false},
		{"III", false},
		{"John", false},
		{"", false},
	}

	for _, tt := range tests {
		t.Run(tt.word, func(t *testing.T) {
			assert.Equal(t, tt.expected, isAcademicSuffix(tt.word))
		})
	}
}

func TestIsParticle(t *testing.T) {
	tests := []struct {
		word     string
		expected bool
	}{
		{"van", true},
		{"Van", true},
		{"VAN", true},
		{"von", true},
		{"de", true},
		{"da", true},
		{"del", true},
		{"della", true},
		{"la", true},
		{"le", true},
		{"bin", true},
		{"ibn", true},
		{"John", false},
		{"", false},
	}

	for _, tt := range tests {
		t.Run(tt.word, func(t *testing.T) {
			assert.Equal(t, tt.expected, isParticle(tt.word))
		})
	}
}

File: app/utils/sortname.test.ts
import { forPerson, forTitle } from "./sortname";
import { describe, expect, it } from "vitest";

describe("forTitle", () => {
  it("moves 'The' to end", () => {
    expect(forTitle("The Hobbit")).toBe("Hobbit, The");
  });

  it("moves 'A' to end", () => {
    expect(forTitle("A Tale of Two Cities")).toBe("Tale of Two Cities, A");
  });

  it("moves 'An' to end", () => {
    expect(forTitle("An American Tragedy")).toBe("American Tragedy, An");
  });

  it("handles lowercase articles", () => {
    expect(forTitle("the hobbit")).toBe("hobbit, the");
  });

  it("handles uppercase articles", () => {
    expect(forTitle("THE HOBBIT")).toBe("HOBBIT, THE");
  });

  it("returns unchanged when no leading article", () => {
    expect(forTitle("Lord of the Rings")).toBe("Lord of the Rings");
  });

  it("returns unchanged when article in middle only", () => {
    expect(forTitle("Return of the King")).toBe("Return of the King");
  });

  it("handles empty string", () => {
    expect(forTitle("")).toBe("");
  });

  it("handles whitespace only", () => {
    expect(forTitle("   ")).toBe("");
  });

  it("returns single word unchanged", () => {
    expect(forTitle("Dune")).toBe("Dune");
  });

  it("returns 'The' alone unchanged", () => {
    expect(forTitle("The")).toBe("The");
  });

  it("handles real-world example: The Lord of the Rings", () => {
    expect(forTitle("The Lord of the Rings")).toBe("Lord of the Rings, The");
  });

  it("handles real-world example: A Game of Thrones", () => {
    expect(forTitle("A Game of Thrones")).toBe("Game of Thrones, A");
  });
});

describe("forPerson", () => {
  // Basic name inversion
  it("inverts simple two-part name", () => {
    expect(forPerson("Stephen King")).toBe("King, Stephen");
  });

  it("inverts three-part name", () => {
    expect(forPerson("Martin Luther King")).toBe("King, Martin Luther");
  });

  // Generational suffixes (preserved)
  it("preserves Jr. suffix", () => {
    expect(forPerson("Robert Downey Jr.")).toBe("Downey, Robert, Jr.");
  });

  it("preserves Jr without period", () => {
    expect(forPerson("Robert Downey Jr")).toBe("Downey, Robert, Jr");
  });

  it("preserves Sr. suffix", () => {
    expect(forPerson("John Smith Sr.")).toBe("Smith, John, Sr.");
  });

  it("preserves III suffix", () => {
    expect(forPerson("John Smith III")).toBe("Smith, John, III");
  });

  it("handles Martin Luther King Jr.", () => {
    expect(forPerson("Martin Luther King Jr.")).toBe(
      "King, Martin Luther, Jr.",
    );
  });

  // Academic suffixes (stripped)
  it("strips PhD suffix", () => {
    expect(forPerson("Jane Doe PhD")).toBe("Doe, Jane");
  });

  it("strips Ph.D. suffix", () => {
    expect(forPerson("Jane Doe Ph.D.")).toBe("Doe, Jane");
  });

  it("strips MD suffix", () => {
    expect(forPerson("John Smith MD")).toBe("Smith, John");
  });

  it("strips multiple academic suffixes", () => {
    expect(forPerson("John Doe MD PhD")).toBe("Doe, John");
  });

  // Mixed generational and academic
  it("preserves Jr. but strips PhD", () => {
    expect(forPerson("John Smith Jr. PhD")).toBe("Smith, John, Jr.");
  });

  // Prefixes (stripped)
  it("strips Dr. prefix", () => {
    expect(forPerson("Dr. Sarah Connor")).toBe("Connor, Sarah");
  });

  it("strips Dr without period", () => {
    expect(forPerson("Dr Sarah Connor")).toBe("Connor, Sarah");
  });

  it("strips Mr. prefix", () => {
    expect(forPerson("Mr. John Smith")).toBe("Smith, John");
  });

  it("strips Mrs. prefix", () => {
    expect(forPerson("Mrs. Jane Doe")).toBe("Doe, Jane");
  });

  it("strips Prof. prefix", () => {
    expect(forPerson("Prof. Albert Einstein")).toBe("Einstein, Albert");
  });

  it("strips Sir prefix", () => {
    expect(forPerson("Sir Isaac Newton")).toBe("Newton, Isaac");
  });

  // Prefix and suffix combined
  it("strips Dr. and PhD", () => {
    expect(forPerson("Dr. John Smith PhD")).toBe("Smith, John");
  });

  // Particles (moved to end)
  it("handles van Beethoven", () => {
    expect(forPerson("Ludwig van Beethoven")).toBe("Beethoven, Ludwig van");
  });

  it("handles von Neumann", () => {
    expect(forPerson("John von Neumann")).toBe("Neumann, John von");
  });

  it("handles da Vinci", () => {
    expect(forPerson("Leonardo da Vinci")).toBe("Vinci, Leonardo da");
  });

  it("handles de Gaulle", () => {
    expect(forPerson("Charles de Gaulle")).toBe("Gaulle, Charles de");
  });

  it("handles del Toro", () => {
    expect(forPerson("Guillermo del Toro")).toBe("Toro, Guillermo del");
  });

  it("handles van der Waals", () => {
    expect(forPerson("Johannes van der Waals")).toBe("Waals, Johannes van der");
  });

  // Particle with suffix
  it("handles particle with Jr.", () => {
    expect(forPerson("John van Smith Jr.")).toBe("Smith, John van, Jr.");
  });

  // Edge cases
  it("handles empty string", () => {
    expect(forPerson("")).toBe("");
  });

  it("handles whitespace only", () => {
    expect(forPerson("   ")).toBe("");
  });

  it("returns single name unchanged", () => {
    expect(forPerson("Madonna")).toBe("Madonna");
  });

  it("handles single name with whitespace", () => {
    expect(forPerson("  Cher  ")).toBe("Cher");
  });

  // Real-world examples
  it("handles J.R.R. Tolkien", () => {
    expect(forPerson("J.R.R. Tolkien")).toBe("Tolkien, J.R.R.");
  });

  it("handles George R.R. Martin", () => {
    expect(forPerson("George R.R. Martin")).toBe("Martin, George R.R.");
  });

  it("handles H.P. Lovecraft", () => {
    expect(forPerson("H.P. Lovecraft")).toBe("Lovecraft, H.P.");
  });
});
